feat(KFLUXVNGD-454): add oci version tag to apply-mapping

yftacherzog opened the pull request:
To pull a Helm chart, we typically need it to be tagged based on the oci artifact version (rather than by digest or arbitrary tag).

To allow that, enhance apply-mapping task to be able to fetch the version from the artifact metadata and populate a variable that can be later used in RPAs templating.

## Describe your changes

## Relevant Jira

## Checklist before requesting a review
- [ ] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [ ] My commit message includes `Signed-off-by: My name <email>`
- [ ] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)
- [ ] I have run the README.md generator script in `.github/scripts/readme_generator.sh` and verified the results using `.github/scripts/check_readme.sh`

yftacherzog committed:
feat(KFLUXVNGD-454): add oci version tag to apply-mapping

To pull a Helm chart, we typically need it to be tagged based on the
oci artifact version (rather than by digest or arbitrary tag).

To allow that, enhance apply-mapping task to be able to fetch the
version from the artifact metadata and populate a variable that
can be later used in RPAs templating.

Signed-off-by: Yftach Herzog <yherzog@redhat.com>

jinqi7 commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -394,10 +396,29 @@ spec:
+            "${arch}" docker://"${IMAGE_REF}" 2>/dev/null | jq -c)" && [[ -n "$image_metadata" ]]; then
+                # Successfully got image metadata - this is a regular container image
+                build_date="$(jq -r '.Labels."build-date" // .Created // ""' <<< "$image_metadata")"
+                oci_version="$(jq -r '.annotations."org.opencontainers.image.version" // ""' <<< "$image_metadata")"
```
I am not sure if oci_version need to do transformation like this "${oci_version//+/_}" since the annotation name is the same with the one in " skopeo  --raw" command.

yftacherzog replied:
At least for Helm charts, the version that goes inside that field needs to be semver-compatible, and then it can have a value, such as: `"org.opencontainers.image.version": "0.1.163+307bbec"`.

When Helm pushes a chart with such version as an OCI artifact, it translates the `+` in the version with underscore for the tag it generates. When it pulls a chart with such version, it will again expect a plus on the version to be equivalent to an underscore on the tag:

```
skopeo inspect --raw docker://quay.io/yftacherzog-konflux/user-ns2/caching/release/squid-helm:0.1.163_307bbec | jq -r '.annotations."org.opencontainers.image.version"'
0.1.163+307bbec

helm pull oci://quay.io/yftacherzog-konflux/user-ns2/caching/release/squid-helm --version 0.1.163+307bbec
Pulled: quay.io/yftacherzog-konflux/user-ns2/caching/release/squid-helm:0.1.163_307bbec
Digest: sha256:60b72337b3a842689d5cd4716122545c1cc509e1afd463efcc45d42090da7b4c
quay.io/yftacherzog-konflux/user-ns2/caching/release/squid-helm:0.1.163_307bbec contains an underscore.

OCI artifact references (e.g. tags) do not support the plus sign (+). To support
storing semantic versions, Helm adopts the convention of changing plus (+) to
an underscore (_) in chart version tags when pushing to a registry and back to
a plus (+) when pulling from a registry.
```

jinqi7 replied:
Does the transformation need to be added in this branch since it is done in the "else" branch in Line#415?

yftacherzog replied:
Since the value is ultimately intended to be used for tagging oci artifacts, I think it will only be usable with that transformation, regardless of the branch in the code. 

If you prefer, maybe we can rename the field to be oci_version_tag or oci_version_sanitized. This way, if in the future there will be a requirement for also delivering the original value for this annotation in RPAs, we can then add an oci_version once again that will provide the non-transformed value. 

Let me know your thoughts.

jinqi7 replied:
If the non-transformed value is/will not be used in other place, it’s fine to go with the name `oci_version`. However, it would be helpful to add a comment on this line explaining that the value’s format has already been transformed.

yftacherzog replied:
I'll change it so that the output is uniform for the two code branches, and I'll add a comment as you suggested.

yftacherzog replied:
@jinqi7 I unified the format for both scenarios and the line in which the transformation is made contains a comment about the transformation. Have a look.

jinqi7 replied:
Looks good about the transformation. But if "skopeo --raw" cmd fails,  the value of "oci_version" will be "".  Although there is an error message, the process will continue.  Is it really expected to release with a wrong tag?

yftacherzog replied:
My concern is that there are variants of non-image artifacts that we might not have in mind at the moment, and for which this value might not be collected for some reason.
I'd prefer releasing a chart with missing tag, over breaking another artifact release process.

jinqi7 replied:
For the annotation, it may be uncertain.  But it may need to exit when "skopeo --raw" command fails.  Let's see any opinion from @mmalina or @johnbieren.

mmalina replied:
It sounds potentially dangerous to assume that just because skopeo failed, it's because it's a non-image artifact. 

skopeo will print this:
```
FATA[0001] unsupported image-specific operation on artifact with type "application/vnd.cncf.helm.config.v1+json"
```
So in theory we could parse this error message and act on that. But parsing error messages also doesn't sound great.

Before this skopeo call, we already call get-image-architectures and it will have already determined that this is not a normal image. So how about we alter get-image-architectures to also include the type in the response? Then we can just decide based on that here.

I checked all the places that call get-image-architecture and they always check specific fields. So we could easily add a new field in that list entry that we return. WDYT?

yftacherzog replied:
I'll look into that.

yftacherzog replied:
@mmalina thinking about what you're saying, are you looking for something like this?

```
if type is not helm (or some other specific things we may come up with in the future):
  do what we were doing up until today + set oci_version
else
  get raw output and process it (as in the else statement in the PR)
```

Let me know if that is what you have in mind.

mmalina replied:
Yeah, except I really meant the opposite - if in get-image-architectures we determine it's a non-image artifact, do this other thing (skopeo inspect --raw), otherwise do as we did until now. Because I think that it's the same reason why for those other types the bash script switched to oras - because skopeo inspect would fail on them.

But: 

> + set oci_version

Why would we want this for normal images? I thought this was only specific to helm artifacts, no? Would anybody want to ever use this with normal images?

yftacherzog replied:
@mmalina, something like this?

```bash
            # Get configMediaType from architecture info to determine artifact type
            config_media_type="$(jq -rs 'map(.configMediaType) | .[0]' <<< "$arch_json")"

            # Check if this is NOT a Helm chart - handle regular container images first
            if [[ "$config_media_type" != "application/vnd.cncf.helm.config.v1+json" ]]; then
                # This is a regular container image - use standard skopeo inspect
                image_metadata="$(skopeo inspect --retry-times 3 --no-tags --override-os "${os}" --override-arch "${arch}" \
                  docker://"${IMAGE_REF}" | jq -c)"
                # For timestamp, use Labels.build-date and fallback to Created
                build_date="$(jq -r '.Labels."build-date" // .Created // ""' <<< "$image_metadata")"
                oci_version_raw="$(jq -r '.annotations."org.opencontainers.image.version" // ""' <<< "$image_metadata")"
            else
                # This is a Helm chart - use raw manifest approach
                echo "Detected Helm chart artifact for ${NAME}, using raw manifest"
                raw_manifest="$(skopeo inspect --retry-times 3 --no-tags --raw docker://"${IMAGE_REF}" | jq -c)"
                oci_version_raw="$(jq -r '.annotations."org.opencontainers.image.version" // ""' <<< "$raw_manifest")"
                build_date="$(jq -r '.annotations."org.opencontainers.image.created" // ""' <<< "$raw_manifest")"
                # Create minimal image_metadata for compatibility (empty if parsing fails)
                image_metadata="$(jq -n --argjson annotations "$(jq '.annotations // {}' <<< "$raw_manifest")" \
                  '{"Labels": {}, "Created": "", "annotations": $annotations}' || echo '{}')"
            fi

            # Transform version to OCI tag format: replace + with _ (convention for OCI compliance)
            # Set default value if empty (common for regular container images without OCI annotations)
            oci_version="${oci_version_raw//+/_}"
            oci_version="${oci_version:-unknown}"
```

yftacherzog replied:
(Looking now at your last message)

yftacherzog replied:
> Why would we want this for normal images? I thought this was only specific to helm artifacts, no? Would anybody want to ever use this with normal images?

Not sure what people may want. What would you prefer the behavior to be when it's not Helm?

yftacherzog replied:
> Yeah, except I really meant the opposite - if in get-image-architectures we determine it's a non-image artifact, do this other thing (skopeo inspect --raw), otherwise do as we did until now. Because I think that it's the same reason why for those other types the bash script switched to oras - because skopeo inspect would fail on them.

I have some reservations about changing that, as it will mean potentially changing the behavior for other non-image artifacts. Up until now, those may relied on the get-image-architectures not returning stuff, and now it will return something. I'm not sure if that will break something somewhere.

mmalina replied:
I was just saying that you're adding the oci version specifically for your help use case and I never heard of it until now, so I am not sure why we would support it for normal images. But I guess it won't hurt and will make it consistent - that way we don't have to explain that it's only possible for some use case. 

Otherwise, yeah, for the most part - I would simplify the image_metadata stuff a bit - it's only used for the labels, so I would deal with that directly - try to get the labels and if not possible, set it to an empty array.

mmalina replied:
> I have some reservations about changing that, as it will mean potentially changing the behavior for other non-image artifacts. Up until now, those may relied on the get-image-architectures not returning stuff, and now it will return something. I'm not sure if that will break something somewhere.

OK, sure, we can have this new path work for the helm type only.

yftacherzog replied:
> OK, sure, we can have this new path work for the helm type only.

New path in get-image-architectures or in apply-mapping?

mmalina replied:
Here. I thought that's what you meant? (But first get-image-architecture will have to be modified to save the type in the output.)

yftacherzog replied:
@mmalina Yes. I just got a little confused :)
Just to see that we're on the same page, you're saying we should go with something similar to the code provided before (in the quote below), given that we enhance get-image-architectures to extract config.mediaType. 

Is that right?

> @mmalina, something like this?
> 
> ```shell
>             # Get configMediaType from architecture info to determine artifact type
>             config_media_type="$(jq -rs 'map(.configMediaType) | .[0]' <<< "$arch_json")"
> 
>             # Check if this is NOT a Helm chart - handle regular container images first
>             if [[ "$config_media_type" != "application/vnd.cncf.helm.config.v1+json" ]]; then
>                 # This is a regular container image - use standard skopeo inspect
>                 image_metadata="$(skopeo inspect --retry-times 3 --no-tags --override-os "${os}" --override-arch "${arch}" \
>                   docker://"${IMAGE_REF}" | jq -c)"
>                 # For timestamp, use Labels.build-date and fallback to Created
>                 build_date="$(jq -r '.Labels."build-date" // .Created // ""' <<< "$image_metadata")"
>                 oci_version_raw="$(jq -r '.annotations."org.opencontainers.image.version" // ""' <<< "$image_metadata")"
>             else
>                 # This is a Helm chart - use raw manifest approach
>                 echo "Detected Helm chart artifact for ${NAME}, using raw manifest"
>                 raw_manifest="$(skopeo inspect --retry-times 3 --no-tags --raw docker://"${IMAGE_REF}" | jq -c)"
>                 oci_version_raw="$(jq -r '.annotations."org.opencontainers.image.version" // ""' <<< "$raw_manifest")"
>                 build_date="$(jq -r '.annotations."org.opencontainers.image.created" // ""' <<< "$raw_manifest")"
>                 # Create minimal image_metadata for compatibility (empty if parsing fails)
>                 image_metadata="$(jq -n --argjson annotations "$(jq '.annotations // {}' <<< "$raw_manifest")" \
>                   '{"Labels": {}, "Created": "", "annotations": $annotations}' || echo '{}')"
>             fi
> 
>             # Transform version to OCI tag format: replace + with _ (convention for OCI compliance)
>             # Set default value if empty (common for regular container images without OCI annotations)
>             oci_version="${oci_version_raw//+/_}"
>             oci_version="${oci_version:-unknown}"
> ```

mmalina replied:
Yes. But I would probably flip the condition to make it more natural:
```
if type = helm
  do this special thing
else
  do the normal thing
```

Also, there is no need for a map here:
```
config_media_type="$(jq -rs '.[0].configMediaType' <<< "$arch_json")"
```

yftacherzog replied:
Alright. I'll work on the change to the script, and then we'll go back here to change the implementation accordingly.

yftacherzog replied:
@mmalina @jinqi7 see https://github.com/konflux-ci/release-service-utils/pull/508

yftacherzog replied:
@mmalina the change is done. I also added a commit here for updating the utils image.

yftacherzog committed:
chore(KFLUXVNGD-454): update utils image reference

The get-image-architectures script was enhanced to output
config.mediaType. Updating references to the utils image.

Signed-off-by: Yftach Herzog <yherzog@redhat.com>

mmalina commented on tasks/managed/apply-mapping/apply-mapping.yaml:
```diff
@@ -370,10 +372,34 @@ spec:
+                oci_version_raw="$(jq -r '.annotations."org.opencontainers.image.version" // ""' <<< "$raw_manifest")"
+                build_date="$(jq -r '.annotations."org.opencontainers.image.created" // ""' <<< "$raw_manifest")"
+                # Create minimal image_metadata for compatibility (empty if parsing fails)
+                image_metadata="$(jq -n --argjson annotations "$(jq '.annotations // {}' <<< "$raw_manifest")" \
```
As I mentioned earlier, `image_metadata` is really only used for labels later. So I would suggest moving the original labels assignment to the else branch below:
```
labels="$(jq -c '.Labels' <<< "${image_metadata}")"
```
And here just `labels="{}"`.

And then `image_metadata` isn't needed in this branch at all. WDYT?

yftacherzog replied:
@mmalina I made the change. See if this is what you had in mind.

mmalina replied:
Yes, that's exactly what I meant, thanks!
